In [1]:
import pandas as pd
import numpy as np

In [2]:
food_reviews = pd.read_csv('food_reviews.csv', encoding="ISO-8859-1")

In [3]:
food_reviews.tail()

product/productId   review/userId       review/profileName  \
568457        B003S1WTCU  A3I8AFVPEE8KI5                R. Sawyer   
568458        B004I613EE  A121AA1GQV751Z            pksd "pk_007"   
568459        B004I613EE   A3IBEVCTXKNOH  Kathy A. Welch "katwel"   
568460        B001LR2CU2  A3LGQPJCZVL9UC                 srfell17   
568461               NaN             NaN                      NaN   

       review/helpfulness  review/score   review/time  \
568457                0/0           2.0  1.331251e+09   
568458                2/2           5.0  1.329782e+09   
568459                1/1           5.0  1.331597e+09   
568460                0/0           5.0  1.338422e+09   
568461                NaN           NaN           NaN   

                            review/summary  \
568457                        disappointed   
568458            Perfect for our maltipoo   
568459  Favorite Training and reward treat   
568460                         Great Honey   
568461                                 NaN   

                                              review/text  
568457  I'm disappointed with the flavor. The chocolat...  
568458  These stars are small, so you can give 10-15 o...  
568459  These are the BEST treats for training and rew...  
568460  I am very satisfied ,product is as advertised,...  
568461                                                NaN

As last row is empty, removing it

In [4]:
food_reviews = food_reviews.iloc[0:len(food_reviews) - 1,]

Checking if the dataframe has empty values

In [6]:
food_reviews.isnull().values.any()

True

It has na values... Finding out which column has them

In [8]:
for i in range(len(food_reviews.columns)):
    print('Column ' + str(i))
    print(food_reviews.iloc[:,i].isnull().values.any())

Column 0
True
Column 1
True
Column 2
True
Column 3
True
Column 4
True
Column 5
True
Column 6
True
Column 7
True


Summary and text columns have missing values. Dropping those rows which have both of them missing.

In [5]:
food_reviews = food_reviews[food_reviews['review/summary'].notnull() | food_reviews['review/text'].notnull()]
food_reviews.tail()

product/productId   review/userId       review/profileName  \
568456        B001EO7N10  A28KG5XORO54AY         Lettie D. Carter   
568457        B003S1WTCU  A3I8AFVPEE8KI5                R. Sawyer   
568458        B004I613EE  A121AA1GQV751Z            pksd "pk_007"   
568459        B004I613EE   A3IBEVCTXKNOH  Kathy A. Welch "katwel"   
568460        B001LR2CU2  A3LGQPJCZVL9UC                 srfell17   

       review/helpfulness  review/score   review/time  \
568456                0/0           5.0  1.299629e+09   
568457                0/0           2.0  1.331251e+09   
568458                2/2           5.0  1.329782e+09   
568459                1/1           5.0  1.331597e+09   
568460                0/0           5.0  1.338422e+09   

                            review/summary  \
568456                 Will not do without   
568457                        disappointed   
568458            Perfect for our maltipoo   
568459  Favorite Training and reward treat   
568460                         Great Honey   

                                              review/text  
568456  Great for sesame chicken..this is a good if no...  
568457  I'm disappointed with the flavor. The chocolat...  
568458  These stars are small, so you can give 10-15 o...  
568459  These are the BEST treats for training and rew...  
568460  I am very satisfied ,product is as advertised,...

Dataframe now has one row less

In [210]:
food_reviews['review/text'].isnull().values.any()

False

In [6]:
food_reviews.index = range(food_reviews.shape[0])

In [7]:
null_summary =  food_reviews['review/summary'][food_reviews['review/summary'].isnull()]

In [238]:
null_summary.index 

Int64Index([ 33958,  40548, 101106, 102979, 117515, 155712, 178290, 198474,
            212691, 237565, 293906, 299495, 300961, 333556, 352043, 357215,
            357814, 360782, 379473, 380558, 381313, 386283, 392529, 484367,
            486640, 503260, 530716],
           dtype='int64')

In [8]:
food_reviews = food_reviews.drop(null_summary.index)

In [9]:
food_reviews['review/summary'][food_reviews['review/summary'].isnull()]

Series([], Name: review/summary, dtype: object)

In [9]:
food_reviews['review/summary'].isnull().values.any()

False

Now the text and summary column don't have any missing value

#### Cleaning the column names

In [9]:
food_reviews.columns = [each.split("/")[1] for each in food_reviews.columns]
len(food_reviews.productId.value_counts()[food_reviews.productId.value_counts() == 1])

30407

In [248]:
food_reviews.shape

(568427, 8)

In [15]:
food_reviews.productId.isnull().sum()

7

In [10]:
food_reviews.loc[:,'helpful_upvotes'] = food_reviews.helpfulness.apply(lambda x: float(x.split("/")[0]))
food_reviews.loc[:,'helpful_responses'] = food_reviews.helpfulness.apply(lambda x: float(x.split("/")[1]))

In [8]:
food_reviews.head()

productId          userId                      profileName helpfulness  \
0  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian         1/1   
1  B00813GRG4  A1D87F6ZCVE5NK                           dll pa         0/0   
2  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"         1/1   
3  B000UA0QIQ  A395BORC6FGVXV                             Karl         3/3   
4  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"         0/0   

   score          time                summary  \
0    5.0  1.303862e+09  Good Quality Dog Food   
1    1.0  1.346976e+09      Not as Advertised   
2    4.0  1.219018e+09  "Delight" says it all   
3    2.0  1.307923e+09         Cough Medicine   
4    5.0  1.350778e+09            Great taffy   

                                                text  helpful_upvotes  \
0  I have bought several of the Vitality canned d...              1.0   
1  Product arrived labeled as Jumbo Salted Peanut...              0.0   
2  This is a confection that has been around a fe...              1.0   
3  If you are looking for the secret ingredient i...              3.0   
4  Great taffy at a great price.  There was a wid...              0.0   

   helpful_responses  
0                1.0  
1                0.0  
2                1.0  
3                3.0  
4                0.0

In [11]:
food_reviews = food_reviews[~(food_reviews.helpful_upvotes > food_reviews.helpful_responses)]

In [12]:
food_reviews['helpfulness_percentage']  = food_reviews.helpful_upvotes /food_reviews.helpful_responses 

In [34]:
food_reviews.head()

productId          userId                      profileName helpfulness  \
0  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian         1/1   
1  B00813GRG4  A1D87F6ZCVE5NK                           dll pa         0/0   
2  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"         1/1   
3  B000UA0QIQ  A395BORC6FGVXV                             Karl         3/3   
4  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"         0/0   

   score          time                summary  \
0    5.0  1.303862e+09  Good Quality Dog Food   
1    1.0  1.346976e+09      Not as Advertised   
2    4.0  1.219018e+09  "Delight" says it all   
3    2.0  1.307923e+09         Cough Medicine   
4    5.0  1.350778e+09            Great taffy   

                                                text  helpful_upvotes  \
0  I have bought several of the Vitality canned d...              1.0   
1  Product arrived labeled as Jumbo Salted Peanut...              0.0   
2  This is a confection that has been around a fe...              1.0   
3  If you are looking for the secret ingredient i...              3.0   
4  Great taffy at a great price.  There was a wid...              0.0   

   helpful_responses  helpfulness_percentage  
0                1.0                     1.0  
1                0.0                     NaN  
2                1.0                     1.0  
3                3.0                     1.0  
4                0.0                     NaN

In [14]:
food_reviews[food_reviews.helpful_upvotes > food_reviews.helpful_responses]

Empty DataFrame
Columns: [productId, userId, profileName, helpfulness, score, time, summary, text, helpful_upvotes, helpful_responses, helpfulness_percentage]
Index: []

### Assigning all nan to -1

In [13]:
food_reviews.helpfulness_percentage[food_reviews['helpful_responses'] == 0] = -1

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [11]:
food_reviews.head()

productId          userId                      profileName helpfulness  \
0  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian         1/1   
1  B00813GRG4  A1D87F6ZCVE5NK                           dll pa         0/0   
2  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"         1/1   
3  B000UA0QIQ  A395BORC6FGVXV                             Karl         3/3   
4  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"         0/0   

   score          time                summary  \
0    5.0  1.303862e+09  Good Quality Dog Food   
1    1.0  1.346976e+09      Not as Advertised   
2    4.0  1.219018e+09  "Delight" says it all   
3    2.0  1.307923e+09         Cough Medicine   
4    5.0  1.350778e+09            Great taffy   

                                                text  helpful_upvotes  \
0  I have bought several of the Vitality canned d...              1.0   
1  Product arrived labeled as Jumbo Salted Peanut...              0.0   
2  This is a confection that has been around a fe...              1.0   
3  If you are looking for the secret ingredient i...              3.0   
4  Great taffy at a great price.  There was a wid...              0.0   

   helpful_responses  helpfulness_percentage  
0                1.0                     1.0  
1                0.0                    -1.0  
2                1.0                     1.0  
3                3.0                     1.0  
4                0.0                    -1.0

### Remove the scores with 3 rating

In [14]:
food_reviews = food_reviews[~(food_reviews['score'] == 3.0)]

In [254]:
food_reviews['score'].value_counts()

5.0    363121
4.0     80654
1.0     52268
2.0     29744
Name: score, dtype: int64

### Giving scores of 4 and above a 1 and a 0 to 2 and below

In [15]:
food_reviews['sentiment'] = food_reviews['score'].apply(lambda x: 0 if x == 1 or x == 2 else 1)

### Making an extended review column by concatinating summary and text.

In [16]:
food_reviews['extended_review'] = food_reviews['summary'] + ' ' + food_reviews['text']

In [40]:
food_reviews.head()

productId          userId                      profileName helpfulness  \
0  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian         1/1   
1  B00813GRG4  A1D87F6ZCVE5NK                           dll pa         0/0   
2  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"         1/1   
3  B000UA0QIQ  A395BORC6FGVXV                             Karl         3/3   
4  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"         0/0   

   score          time                summary  \
0    5.0  1.303862e+09  Good Quality Dog Food   
1    1.0  1.346976e+09      Not as Advertised   
2    4.0  1.219018e+09  "Delight" says it all   
3    2.0  1.307923e+09         Cough Medicine   
4    5.0  1.350778e+09            Great taffy   

                                                text  helpful_upvotes  \
0  I have bought several of the Vitality canned d...              1.0   
1  Product arrived labeled as Jumbo Salted Peanut...              0.0   
2  This is a confection that has been around a fe...              1.0   
3  If you are looking for the secret ingredient i...              3.0   
4  Great taffy at a great price.  There was a wid...              0.0   

   helpful_responses  helpfulness_percentage  sentiment  \
0                1.0                     1.0          1   
1                0.0                    -1.0          0   
2                1.0                     1.0          1   
3                3.0                     1.0          0   
4                0.0                    -1.0          1   

                                     extended_review  
0  Good Quality Dog Food I have bought several of...  
1  Not as Advertised Product arrived labeled as J...  
2  "Delight" says it all This is a confection tha...  
3  Cough Medicine If you are looking for the secr...  
4  Great taffy Great taffy at a great price.  The...

### Tokenize the words after converting the text to lower case

In [16]:
from nltk.tokenize import word_tokenize
from gensim.models import Word2Vec

C:\ProgramData\Anaconda3\lib\site-packages\gensim\utils.py:1212: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")


In [17]:
food_reviews['review_tokens'] = food_reviews.text.str.lower().apply(lambda x: word_tokenize(x))

### Remove stop words, stemming, and punctuations. 

In [17]:
from nltk.stem import PorterStemmer
from nltk.corpus import stopwords

In [19]:
stemmer = PorterStemmer()
food_reviews['review_tokens'] = food_reviews['review_tokens'].apply(lambda x: [stemmer.stem(words) for words in x])

In [18]:
def remove_punctuations_and_stopwords(x):
    #Removing stopwords and punctuations
    y = []
    stop_words = set(stopwords.words('english'))
    for i in range(len(x)):
        if x[i].isalnum() and x[i] not in stop_words:
            y.extend([x[i]])
    return y        

In [22]:
food_reviews['review_tokens'] = food_reviews['review_tokens'].apply(remove_punctuations_and_stopwords)

### Saving word2vec model. Use it load in the future.

In [23]:
model = Word2Vec(food_reviews.review_tokens, size=100, window=5, min_count=1, workers=4)
model.save("word2vec_binary.model")

In [24]:
food_reviews['review_tokens']

0         [bought, sever, vital, dog, food, product, fou...
1         [product, arriv, label, jumbo, salt, peanut, p...
2         [thi, confect, ha, around, centuri, light, pil...
3         [look, secret, ingredi, robitussin, believ, fo...
4         [great, taffi, great, price, wa, wide, assort,...
5         [got, wild, hair, taffi, order, thi, five, pou...
6         [thi, saltwat, taffi, great, flavor, wa, veri,...
7         [thi, taffi, good, veri, soft, chewi, flavor, ...
8         [right, mostli, sprout, thi, cat, eat, grass, ...
9         [thi, veri, healthi, dog, food, good, digest, ...
10        [know, cactu, tequila, uniqu, combin, ingredi,...
11        [one, boy, need, lose, weight, put, thi, food,...
12        [cat, happili, eat, felida, platinum, two, yea...
13        [good, flavor, came, secur, pack, fresh, delic...
14        [strawberri, twizzler, guilti, pleasur, yummi,...
15        [daughter, love, twizzler, thi, shipment, six,...
16        [love, eat, good, watch, tv, l

In [52]:
pd.DataFrame(model[food_reviews.review_tokens[0]])

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


0         1         2         3         4         5         6   \
0   1.946060  1.860980  0.532673  1.625578 -3.681472 -1.084165  0.112747   
1   1.578583 -0.422060 -2.591665  1.522952  0.305976 -2.303166  0.417739   
2   2.022123  0.903349  0.870929 -2.112335  1.658859 -0.577002 -0.751362   
3   1.366296 -1.479822 -0.413918  0.561129 -2.302040  3.493908  1.738750   
4   1.624344  1.572355 -2.075810 -0.288095 -0.941047 -1.406459  0.017346   
5  -0.568345 -0.659343 -0.288215 -2.017586 -0.790885  1.129868 -0.753193   
6   0.381403  1.185182  0.860117  2.182082 -1.743337 -0.343917 -1.001930   
7  -0.232182  0.619505  2.445127  0.116381  0.286503 -0.196643 -0.737453   
8  -1.752139 -0.517439 -0.760262  0.426504  0.041284  2.369285  1.970862   
9  -0.568345 -0.659343 -0.288215 -2.017586 -0.790885  1.129868 -0.753193   
10 -0.143357  1.083417 -1.890144  3.022065 -0.767070 -0.012641  0.376071   
11  1.425535  1.242238  0.498384 -0.398907 -0.374851  0.689526  1.357635   
12  0.828179  2.489445 -1.596724 -1.058282  1.936712  0.919189  0.842728   
13  1.521547  0.434549 -0.766739 -0.448283 -0.451195  1.696368 -2.568662   
14  0.847204 -0.319420 -0.602030 -0.633178  1.486239  1.304646 -0.208212   
15 -0.236355 -0.825386 -1.249998  1.532462 -0.374377  5.230303  3.944634   
16  0.441731  1.172604  1.938308  0.696939 -1.274239  0.750119  0.054163   
17 -0.001953 -1.298455  1.048871 -0.710167  0.137651 -1.215120  0.716963   
18 -0.575650 -1.462667  0.711971 -0.891491 -0.908531 -1.315572  1.029927   
19  0.326650  0.712215  0.106168 -0.885261 -0.387444  0.401051  0.463965   
20 -0.669093  0.530090 -0.336361  0.772716  0.208703  1.580499  3.137695   
21 -0.568345 -0.659343 -0.288215 -2.017586 -0.790885  1.129868 -0.753193   
22  0.441731  1.172604  1.938308  0.696939 -1.274239  0.750119  0.054163   

          7         8         9     ...           90        91        92  \
0   0.676356  1.351604 -0.519378    ...     1.081334  1.968274 -1.662917   
1  -0.678996  0.481408  0.058045    ...     2.945388 -0.872796 -1.207488   
2   0.663725  0.292268  0.275667    ...     0.420612 -0.353272  1.916561   
3   0.261763 -0.038713 -2.483709    ...    -1.700613 -3.331706  3.159449   
4  -0.018068 -1.876335 -0.756309    ...    -0.303187  0.144177  0.727231   
5  -0.983217 -0.386012  0.637971    ...    -0.131770  1.415093  1.324208   
6  -0.235982  0.076646 -2.742886    ...    -0.378958  2.110864 -2.083697   
7   0.981102  0.523327 -0.264144    ...     2.331611 -0.338675 -0.448713   
8  -1.938668  1.550547 -0.763631    ...    -0.521572  2.371176  1.545478   
9  -0.983217 -0.386012  0.637971    ...    -0.131770  1.415093  1.324208   
10  0.985505 -1.013091 -1.197528    ...    -0.094100  1.449169  3.127985   
11 -0.483272  0.361451 -0.211070    ...     0.362192  0.208181 -2.143730   
12 -1.224775  1.930424 -0.650474    ...     1.731965 -0.737943  1.592689   
13  0.412148 -2.153170  0.002422    ...    -1.020719 -0.936207  1.172981   
14 -0.334462 -1.287776 -0.254229    ...     0.838243 -1.925654  3.683582   
15 -5.904045 -0.786353 -0.733844    ...    -2.707828 -2.076708 -1.816662   
16  1.407175 -1.333194  1.209635    ...     1.423175 -0.258233 -1.298598   
17 -0.174637  1.193595 -0.884512    ...    -1.489460  0.096678 -0.459114   
18  0.371725 -2.101332 -1.978958    ...    -0.164339 -1.859780  1.852857   
19  0.768959 -0.186793  0.581203    ...     1.517630  0.903830  2.370842   
20  0.017341  0.706602 -0.972045    ...    -0.945143 -0.452105  0.884598   
21 -0.983217 -0.386012  0.637971    ...    -0.131770  1.415093  1.324208   
22  1.407175 -1.333194  1.209635    ...     1.423175 -0.258233 -1.298598   

          93        94        95        96        97        98        99  
0  -2.862293 -0.276163 -0.722744 -2.260351  0.112261 -0.123738  1.575836  
1  -0.004340 -0.216477  1.353301  0.445696 -1.664177  0.905811  1.527169  
2  -0.764074 -0.641312 -0.361011  0.819080  1.207019 -0.324846  0.121718  
3  -0.828991 -5.540031 -1.041232  0.643751 -0.024014  0.4003

In [53]:
def get_mean(x):
    # Gets the mean of the sentence so that the sentence is represented in the 100-d space
    x = pd.DataFrame(x)
    return x.mean()

def get_sum(x):
    # Gets the sum of the sentence so that the sentence is represented in the 100-d space
    x = pd.DataFrame(x)
    return x.sum()
    

In [95]:
food_reviews.index = range(food_reviews.shape[0])

In [97]:
df_model_avg = []
for i in range(0, len(food_reviews)):
    df_model_avg.append(get_mean(model[food_reviews.review_tokens[i]]))
    
df_model_avg = pd.DataFrame(df_model_avg)    

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  This is separate from the ipykernel package so we can avoid doing imports until


In [98]:
df_model_sum = []
for i in range(0, len(food_reviews)):
    df_model_sum.append(get_sum(model[food_reviews.review_tokens[i]]))
    
df_model_sum = pd.DataFrame(df_model_sum)    

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  This is separate from the ipykernel package so we can avoid doing imports until


### Performing logistic regression and cross- validation to get the sentiment

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics, cross_validation

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [101]:
predicted = cross_validation.cross_val_predict(LogisticRegression(), df_model_avg, food_reviews['sentiment'], cv=10)
print(metrics.accuracy_score(food_reviews['sentiment'], predicted))

0.9063543623956851


Accuracy of this is 90.6%

In [102]:
predicted = cross_validation.cross_val_predict(LogisticRegression(), df_model_sum, food_reviews['sentiment'], cv=10)
print(metrics.accuracy_score(food_reviews['sentiment'], predicted))

0.90180901158589


Accuracy of this is 90.18%. It decreased very slightly. Going with the mean model only

### Implementing the same using TF-IDF

In [19]:
from nltk import FreqDist

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

In [21]:
food_reviews['extended_review_processed'] = food_reviews['extended_review'].apply(lambda x: " ".join(x.lower() for x in x.split()))
stop_words = set(stopwords.words('english'))
food_reviews['extended_review_processed'] = food_reviews['extended_review'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_words))


In [22]:
stemmer = PorterStemmer()
food_reviews['extended_review_processed'] = food_reviews['extended_review_processed'].apply(lambda x: [stemmer.stem(words) for words in x.split(' ')])

In [23]:
food_reviews['extended_review_processed'] = food_reviews['extended_review_processed'].apply(lambda x : ' '.join(x))

In [24]:
vectorizer = TfidfVectorizer(max_df = 0.8, max_features=3000)
tfidf = vectorizer.fit_transform(food_reviews['extended_review_processed'])


In [305]:
model = None
del model

In [25]:
feature_names = vectorizer.get_feature_names()
corpus_index = food_reviews.index
tfidf_df = pd.DataFrame(tfidf.todense(), columns=feature_names)
tfidf_df.head()

00   10  100   11   12  120   13   14  140   15  ...    yr  yuck  yum  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0  0.0   

      yummi  yummy  zero  zevia  zico  zip  zuke  
0  0.000000    0.0   0.0    0.0   0.0  0.0   0.0  
1  0.000000    0.0   0.0    0.0   0.0  0.0   0.0  
2  0.149338    0.0   0.0    0.0   0.0  0.0   0.0  
3  0.000000    0.0   0.0    0.0   0.0  0.0   0.0  
4  0.286871    0.0   0.0    0.0   0.0  0.0   0.0  

[5 rows x 3000 columns]

### Performing logistic regression and 10 fold cross- validation to get the sentiment

In [26]:
tfidf.shape


(525787, 3000)

In [29]:
predicted = cross_validation.cross_val_predict(LogisticRegression(), tfidf, food_reviews['sentiment'], cv=10)
print(metrics.accuracy_score(food_reviews['sentiment'], predicted))

0.940595716516384


The TF-IDF method gives an accuracy of 94%